### PR TITLE
Support rgb colors with percentage

### LINKED ISSUE #150: RGB percentage syntax not supported
The SVG standard allows a color to be specified by the hex values e.g. `#ff00ff` as well as the RGB value `rgb(255, 0, 255)`. For the latter, percentage values are also supported e.g. `rgb(100%, 0%, 100%)`. SNV doesn't seem to support the percentage values.

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1000 1000" width="1000" height="1000">
  <rect x="100" y="100" width="200" height="200" fill="rgb(255, 255, 0)" />
  <rect x="400" y="100" width="200" height="200" fill="rgb(100%, 100%, 0%)" />
</svg>
```
Rendering by Librsvg:
image
Rendering by SNV:
image

### PR DESCRIPTION
**Pull request by moazin**
Fixes #150. @dirkschulze Please let me know if you want me to code this differently.

### CODE REVIEW

File: svgnative/src/SVGStringParser.cpp [hunk 1]
```diff
@@ -712,7 +730,7 @@ void ParsePathString(const std::string& pathString, Path& p)
                 currentY += newY;
             }
 
-            SVG_ASSERT(angle == 0);
+            //SVG_ASSERT(angle == 0);
```
Comment (dirkschulze) [inline svgnative/src/SVGStringParser.cpp:733]: We should remove it entirely.

File: svgnative/src/SVGStringParser.cpp [hunk 2]
```diff
@@ -936,25 +954,37 @@ static bool ParseColor(CharIt& pos, const CharIt& end, ColorImpl& paint, bool su
             std::int32_t r{};
             std::int32_t g{};
             std::int32_t b{};
+            bool r_percentage = false;
```
Comment (dirkschulze) [inline svgnative/src/SVGStringParser.cpp:957]: The rest of the code uses camel case. Also, nit, use intialization lists.
```cpp
bool rPercentage{};
```

Comment (dirkschulze) [reply svgnative/src/SVGStringParser.cpp:957]: Oh, another nit, boolean values should be descriptive:

`hasRPercentage` or `isRPrercentage`... depending on the context.

File: svgnative/src/SVGStringParser.cpp [hunk 3]
```diff
@@ -936,25 +954,37 @@ static bool ParseColor(CharIt& pos, const CharIt& end, ColorImpl& paint, bool su
...
             if (!SkipOptWspDelimiterOptWsp(pos, end))
                 return false;
-            if (!ParseDigit(pos, end, b))
+            if (!ParseDigitOrPercentage(pos, end, b, b_percentage))
                 return false;
             if (!SkipOptWsp(pos, end))
                 return false;
             if (*pos++ != ')')
                 return false;
-            color[0] = std::min(255, r) / 255.0f;
```
Comment (dirkschulze) [inline svgnative/src/SVGStringParser.cpp:955]: I'd suggest something like:

```cpp
// All values must either be digits or percentages.
if (hasRPercentage != hasGPercentage || hasRPercentage  != hasBPercentage)
    return false;
float base = hasRPercentage ? 100.0f : 255.0f;
color[0] = std::min(base, r) / base;
color[1] = std::min(base, g) / base;
color[2] = std::min(base, b) / base;
```

File: svgnative/src/SVGStringParser.cpp (hunk 2 above)
Comment (dirkschulze) [reply svgnative/src/SVGStringParser.cpp:957]:
```cpp
bool hasRPercentage{};
bool hasGPercentage{};
bool hasBPercentage{};
```

File: svgnative/src/SVGStringParser.cpp (hunk 1 above)
Comment (moazin) [reply svgnative/src/SVGStringParser.cpp:733]: Are you sure, I filed an issue #134 that shows problems with arcs, so I think the assertion is there for a good reason. I commented it out by mistake, I'm removing the comment and putting the assertion back for now.

**Comment by moazin** [comment]
I think this needs some more work. I think the RGB syntax does allow for floating point values as well. So for example, `rgb(50.55%, 24.56%, 64.2%)` should be possible too right? The code I wrote only allows for integer percentages.

File: svgnative/src/SVGStringParser.cpp (hunk 1 above)
Comment (dirkschulze) [reply svgnative/src/SVGStringParser.cpp]: Sounds good then.

**Comment by dirkschulze** [comment]
@moazin Yes, you are right, percentage values may take floats. https://drafts.csswg.org/css-color-3/#rgba-color

And yes, we should support it. 2 choices:
1. Open a new issue to support percentage with floats and only support digits for now
2. Do all in the same patch.

**Comment by moazin** [comment]
@dirkschulze Done. I've tested it with a few examples, seems to work well, please review. :)

**Review by dirkschulze — APPROVED** [review]

**Pull request merged**